### PR TITLE
T-35 Add Valuation history info to asset info page

### PR DESCRIPTION
**Pull request by goremukin**

### CODE REVIEW

**Comment by goremukin** [comment]
[T-35 Add Valuation history info to asset info page](https://trello.com/c/uDBil0uT/35-t-35-add-valuation-history-info-to-asset-info-page)

File: PortfolioKeeper.Tests/BusinessLogic/Services/AssetValuationServiceTests.cs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        [Test]
+        public void AssetValuationService_When_UnitOfWork_Null_Shoud_Throws_ArgumentNullException()
+        {
+            Action act = () => new AssetValuationService(null, mapper.Object);
+
+            act.Should().Throw<ArgumentNullException>();
+        }
+
+        [Test]
+        public void AssetValuationService_When_Mapper_Null_Shoud_Throws_ArgumentNullException()
```
Comment (AntonBessarabov) [inline PortfolioKeeper.Tests/BusinessLogic/Services/AssetValuationServiceTests.cs]: Should Throw?

File: PortfolioKeeper.Tests/BusinessLogic/Services/AssetValuationServiceTests.cs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                            IOrderedQueryable<AssetValuation>>>(),
+                        It.IsAny<string>()))
+                .Returns(data);
+            uof.Setup(x => x.GetRepository<AssetValuation, int>()).Returns(assetValuationRepository.Object);
+
+            underTest = new AssetValuationService(uof.Object, mapper.Object);
+        }
+
+        [Test]
+        public void AssetValuationService_When_UnitOfWork_Null_Shoud_Throws_ArgumentNullException()
```
Comment (AntonBessarabov) [inline PortfolioKeeper.Tests/BusinessLogic/Services/AssetValuationServiceTests.cs]: Should Throw?

File: PortfolioKeeper.Tests/BusinessLogic/Services/AssetValuationServiceTests.cs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        public void Get_When_All_Id_Valid_Should_Call_Map_Method(TimeRange range)
+        {
+            underTest.Get(AssetId, range);
+
+            mapper.Verify(moq =>
+                moq.Map<IEnumerable<AssetValuation>, IEnumerable<AssetValuationDto>>(It.IsAny<IEnumerable<AssetValuation>>()), Times.Once);
+        }
+
+        [Test]
+        public void Get_When_TimeRange_Not_Valid_Should_Throws_ArgumentException()
```
Comment (AntonBessarabov) [inline PortfolioKeeper.Tests/BusinessLogic/Services/AssetValuationServiceTests.cs]: Throw?

File: PortfolioKeeper/App_Start/AutofacConfig.cs [hunk 4]
```diff
@@ -99,6 +99,10 @@ private static void RegisterBusinessComponents(ContainerBuilder builder)
                 .As<IAssetService>()
                 .InstancePerRequest();
 
+            builder.RegisterType<AssetValuationService>()
+                .As<IAssetValuationService>()
+                .InstancePerRequest();
```
Comment (AntonBessarabov) [inline PortfolioKeeper/App_Start/AutofacConfig.cs]: Change to .InstancePerLifetimeScope(), please

File: PortfolioKeeper/BusinessLogic/Services/AssetValuationService.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+            {
+                throw new ArgumentNullException(nameof(unitOfWork));
+            }
+
+            assetValuationRepository = unitOfWork.GetRepository<AssetValuation, int>()
+                                       ?? throw new ArgumentNullException(nameof(assetValuationRepository));
+            this.mapper = mapper ?? throw new ArgumentNullException(nameof(mapper));
+        }
+
+        public IEnumerable<AssetValuationDto> Get(int assetId, TimeRange range = TimeRange.AllTime)
```
Comment (AntonBessarabov) [inline PortfolioKeeper/BusinessLogic/Services/AssetValuationService.cs]: I suggest move arguments to the model

File: PortfolioKeeper/Controllers/AssetController.cs [hunk 6]
```diff
@@ -126,6 +130,46 @@ public ActionResult Details(int id)
...
+                    range = TimeRange.HalfAYear;
+                    break;
+                default:
+                    throw new ArgumentException("Invalid button type!");
+            }
+
+            if (!Enum.IsDefined(typeof(TimeRange), range))
+            {
+                throw new ArgumentException("Invalid range value!");
+            }
```
Comment (AntonBessarabov) [inline PortfolioKeeper/Controllers/AssetController.cs]: I'm not sure that this check is necessary

File: PortfolioKeeper/Views/Asset/Valuations.cshtml [hunk 7]
```diff
@@ -0,0 +1,28 @@
+﻿@model IEnumerable<PortfolioKeeper.Models.ViewModels.AssetValuationViewModel>
+
+<h2>Asset valuations:</h2>
+<br />
+
+<div id="valuations" class="form-group">
```
Comment (AntonBessarabov) [inline PortfolioKeeper/Views/Asset/Valuations.cshtml]: Why you use form-group here?

File: PortfolioKeeper/Views/Shared/_Layout.cshtml [hunk 8]
```diff
@@ -6,6 +6,12 @@
     <title>@ViewBag.Title - My ASP.NET Application</title>
     @Styles.Render("~/Content/css")
     @Scripts.Render("~/bundles/modernizr")
+    
+    @Scripts.Render("~/bundles/jquery")
+    @Scripts.Render("~/bundles/bootstrap")
+    <script src=@Url.Content("~/Scripts/jquery.unobtrusive-ajax.min.js") type="text/javascript"></script>
+    <script src=@Url.Content("~/Scripts/MicrosoftAjax.js") type="text/javascript"></script>  
+    <script src=@Url.Content("~/Scripts/MicrosoftMvcAjax.js") type="text/javascript"></script>
```
Comment (AntonBessarabov) [inline PortfolioKeeper/Views/Shared/_Layout.cshtml]: Move to bundles, please

Comment (AntonBessarabov) [reply PortfolioKeeper/Views/Shared/_Layout.cshtml]: And render before </body>

File: PortfolioKeeper/Views/Asset/Valuations.cshtml [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    @if (!Model.Any())
+    {
+        <label class="col-sm-2 control-label">No data</label>
+        <br/>
+    }
+    else
+    {
+        <table class="table table-bordered table-responsive table-hover">
+            <tr>
```
Comment (EdwOK) [inline PortfolioKeeper/Views/Asset/Valuations.cshtml]: **True** table structure: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/table

File: PortfolioKeeper/Views/Asset/Valuations.cshtml [hunk 10]
```diff
@@ -0,0 +1,28 @@
+﻿@model IEnumerable<PortfolioKeeper.Models.ViewModels.AssetValuationViewModel>
+
+<h2>Asset valuations:</h2>
+<br />
+
+<div id="valuations" class="form-group">
+    @if (!Model.Any())
+    {
+        <label class="col-sm-2 control-label">No data</label>
+        <br/>
```
Comment (EdwOK) [inline PortfolioKeeper/Views/Asset/Valuations.cshtml]: `<br>` - bad style, see bootstrap docs

File: PortfolioKeeper/BusinessLogic/Services/AssetValuationService.cs [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+        {
+            IEnumerable<AssetValuation> valuations;
+
+            if (range == TimeRange.AllTime)
+            {
+                valuations = assetValuationRepository.Get(e => e.AssetId == assetId,
```
Comment (EdwOK) [inline PortfolioKeeper/BusinessLogic/Services/AssetValuationService.cs]: **39** line and **47** line - exp dublicate, _solution_: `Specification pattern` or smth

File: PortfolioKeeper/Controllers/AssetController.cs [hunk 12]
```diff
@@ -126,6 +130,46 @@ public ActionResult Details(int id)
...
+                throw new SecurityException("User claim value not found");
+            }
+
+            securityService.CheckAssetAccess(id, userId);
+
+            TimeRange range;
+
+            switch (button)
+            {
+                case "Week":
```
Comment (EdwOK) [inline PortfolioKeeper/Controllers/AssetController.cs]: `enum keys`

File: PortfolioKeeper/Views/Asset/Valuations.cshtml [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+    }
+    else
+    {
+        <table class="table table-bordered table-responsive table-hover">
```
Comment (EdwOK) [inline PortfolioKeeper/Views/Asset/Valuations.cshtml]: `<br>` - bad style, see bootstrap docs

Comment (EdwOK) [reply PortfolioKeeper/Views/Asset/Valuations.cshtml]: True table structure: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/table

File: PortfolioKeeper/Controllers/AssetController.cs (hunk 6 above)
Comment (goremukin) [reply PortfolioKeeper/Controllers/AssetController.cs]: @AntonBessarabov, Yes. No more

File: PortfolioKeeper/BusinessLogic/Services/AssetValuationService.cs (hunk 11 above)
Comment (goremukin) [reply PortfolioKeeper/BusinessLogic/Services/AssetValuationService.cs]: @EdwOK, better?

            Predicate<DateTime> predicate;

            if (range == TimeRange.AllTime)
            {
                predicate = time => true;
            }
            else
            {
                int days = GetDaysCount(range);
                DateTime start = DateTime.UtcNow.Date.AddDays(-days);
                predicate = time => time >= start;
            }

            valuations = assetValuationRepository
                .Get(e => e.AssetId == assetId && predicate.Invoke(e.CreatedOn),
                    q => q.OrderBy(e => e.CreatedOn));

File: PortfolioKeeper/BusinessLogic/Services/AssetValuationService.cs (hunk 5 above)
Comment (goremukin) [reply PortfolioKeeper/BusinessLogic/Services/AssetValuationService.cs]: @AntonBessarabov, I think, while id + range meets once, it is not necessary to do it

File: PortfolioKeeper/Views/Asset/Valuations.cshtml (hunk 13 above)
Comment (goremukin) [reply PortfolioKeeper/Views/Asset/Valuations.cshtml]: @EdwOK, 👍

File: PortfolioKeeper.Tests/BusinessLogic/Services/AssetValuationServiceTests.cs [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+        [Test]
+        public void AssetValuationService_When_UnitOfWork_Null_Shoud_Throw_ArgumentNullException()
+        {
+            Action act = () => new AssetValuationService(null, mapper.Object);
+
+            act.Should().Throw<ArgumentNullException>();
+        }
+
+        [Test]
+        public void AssetValuationService_When_Mapper_Null_Shoud_Throw_ArgumentNullException()
```
Comment (AntonBessarabov) [inline PortfolioKeeper.Tests/BusinessLogic/Services/AssetValuationServiceTests.cs]: Should

**Pull request merged**